TST: Add base branch check

pllim opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

This pull request is to perform a check when a new PR is opened and fails the check if the PR is not opened against the correct base branch (`master`). This is because occasionally, a contributor would open PR via the "edit on GitHub" link from RTD docs, which results in them editing an older version of the code on the wrong branch. If a PR is opened against a release branch on purpose, then a special label could be applied by maintainer to bypass the check.

**Notes**

- There is no special event for simply changing base branch, so this check might run even when it is not necessary, unless you only want it to run once when PR is opened. (See below.)
- Update: This action now uses Javascript and completes in 2-3s, so having it running when not strictly needed isn't such a big issue anymore. Prior, it took like 20s when running on Docker.

**TODO or Needs Discussions**

- [x] I did not put this in the labeler YAML, as I originally promised, because you probably want this check not just when the PR is opened? But if you only wants to let it run once when opened, then I can move it to labeler YAML. (Looks like people are okay with running this per push event.)
- [ ] When special label is applied, manually re-running the CI will not change the failure to success because the workflow is not watching label/unlabel events and not sure if we want it to. This means, only force pushing existing commit or normal pushing new commits would change the status... ~maintainer needs to manually re-run the CI because labeled/unlabeled does not trigger CI and we don't want it to anyway.~ Is this okay? If not, another option is to move this out to a different YAML. ~This point is moot if we move to labeler YAML.~
- [x] Maybe we can have the rest of CI to only run if base branch check is successful ~but setting up the logic may be tricky because the YAML also handle push and tag events. Tricky logic might not be needed, though, if I modify the Action code to be a successful no-op if event is not from a PR~. (Action modified to no-op when not a PR.) ~But this point is moot if we move to labeler YAML.~
- [x] Add new `skip-basebranch-check` label if we decide to proceed with this. ~But @bsipocz also suggested an alternative approach where the Action slaps a `non-default-basebranch` label instead of failing the check, which is kind of opposite way of using label as it currently is.~ (I added the label but we can remove it if this PR does not get into production.)
- [x] ~Change `pull_request:` to `pull_request_target:` according to best practices from https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_target . But we can only do this after approval but before merge, because otherwise the CI won't run properly here.~ (Not needed anymore here because no longer uses token.)
- [x] Squash commits?
- [x] ~Rename actions to something @astrofrog likes. (Or we can defer?)~ Not applicable anymore.
- [x] ~Move actions to `astropy` org. 🐶 (Or we can defer?)~ Not applicable anymore.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

pllim commented on .github/workflows/ci_workflows.yml:
```diff
@@ -6,13 +6,25 @@ on:
     tags:
     - '*'
-  pull_request:
+  pull_request_target:
```
Oh, weird, the CI didn't kick off. Did this break it? I was just trying to follow best practices stated in https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_target .

embray commented:
LGTM: Is your concern about CI just to not waste time on builds that might not be useful?  Other than that, I think just as an alert that the base branch is not correct is pretty sufficient.

pllim commented:
> Is your concern about CI just to not waste time on builds that might not be useful?

Partly. Another concern is there might be too many checks to the point people stop paying attention.

Brigitta's idea to apply a label is pretty good too, so I guess I am just waiting for a consensus on how to move forward.

bsipocz commented:
after this morning call:
- I think this need to be run once if this is only for labeling. To keep the failing status, I think it should be run for each commit, otherwise new commits will just override the status. This is a quick action, so I don't think it should be an issue if it always runs tbh
- Matt at Pay Lian totally convinced me that setting a status is better than a label, so I agree that the status should be set to fail, unless a magic label or something is done (as we occasionally need these PRs against the non default branch)

pllim commented:
I updated the Action code upstream and this PR to match. I cancelled the rest of the CI because I just want to see if my action would run (it did). I also updated the todo list above to move the discussions along. Thanks for the feedback thus far!

pllim commented:
The event of applying special label to skip the check might need more discussions, because I found that simply manually re-running the workflow after applying the special label to skip check is insufficient, most likely because workflow is not watching label/unlabel event, so that's why its `github.event` context failed to pick up the newly added label?

pllim commented:
Further trimmed down the run time by rewriting this action in Typescript. It now completes in 2-3s.

pllim commented:
Update: I removed the need for custom action using `github-script` at the expense of more LOC here. Small price to pay for one less puppy, I guess.

embray commented on .github/workflows/ci_workflows.yml:
```diff
@@ -17,6 +17,23 @@ jobs:
+      with:
+        script: |
+          const skip_label = 'skip-basebranch-check';
+          const allowed_basebranch = 'master';
```
I don't know much about github-script, but I wonder if there is a way to retrieve the repository's default branch rather than hard-coding it (e.g. if we ever switch to "main").

embray replied:
It looks like

```javascript
const { default_branch } = github.repos.get(context.repo);
```

would do it.

pllim replied:
I don't think this works. It says undefined...

pllim replied:
OK fixed it.

embray replied:
Sorry for the misleading untested suggestion.  The docs on this were...not clear, and from the code it really looked like it should work :/  Thanks for figuring out the right answer.

pllim commented:
@embray , what about now?

embray commented:
Looks like there is a merge conflict now.

embray reviewed (approved):
I fixed the merge conflict (it was caused by #11330).  LGTM now.

pllim commented:
Hold on, don't merge yet. I'll see if I can rebase to get rid of the merge conflict.

pllim commented on .github/workflows/ci_workflows.yml:
```diff
@@ -13,9 +13,32 @@ env:
   tests:
     name: ${{ matrix.name }}
     runs-on: ${{ matrix.os }}
+    needs: initial_checks
```
Had to add this back. It was removed in another PR but the rebase didn't pick it back up.

pllim commented:
@embray , it is ready now. Thanks for the review! Please merge if you are okay with this.

pllim commented:
Since this is approved, let's try it out. Can take it out if this turns out to be more trouble than it's worth.
